parlai/agents/t5/modules.py

```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""
+Wrapped Encoders for ParlAI Use.
+"""
+import torch
+from transformers import T5ForConditionalGeneration
+
+try:
+    from transformers.models.t5.modeling_t5 import T5Stack
+except ModuleNotFoundError:
+    # Prior versions of transformers package do not have T5Stack
+    T5Stack = object
+from typing import Optional, Dict, Any, Tuple
+
+from parlai.core.opt import Opt
+from parlai.core.torch_generator_agent import TorchGeneratorModel
+
+from parlai.agents.t5.dict import T5TokenizerDictionaryAgent
+
+
+def build_t5(opt: Opt) -> T5ForConditionalGeneration:
+    return T5ForConditionalGeneration.from_pretrained(
+        opt['t5_model_arch'], dropout_rate=opt['t5_dropout']
+    )
+
+
+def set_device(func):
+    """
+    Decorator for setting device.
+
+    HF's model parallel uses `torch.cuda.set_device`, which does not vibe well with
+    ParlAI.
```

Comment:
Hmm what happens when trying to use it in ParlAI?

Reply:
HF uses `torch.cuda.set_device()` for model parallelism; this sets the device globally, and HF does not reset the device after model forward calls. So, when ParlAI calls `.cuda()` on the next batch tensors, it'll copy the tensors to whatever device was set by HF in the last forward call --> almost certainly not `cuda:0`, which is where embeddings for the model generally live. so we need to reset the device after every call

I wrap both sides of the function call with a `set_device` command because in my experience the wrapper did not fully execute if there was an OOM